src/transcoding.ts

```diff
@@ -0,0 +1,318 @@
+      );
+    } else {
+      resultList.push(
+        `${prefix}${encodeWithoutSlashes(key)}=${encodeWithoutSlashes(
+          request[key].toString()
+        )}`
+      );
+    }
+  }
+  return resultList;
+}
+
+export function encodeWithSlashes(str: string): string {
+  return str
+    .split('')
+    .map(c => (c.match(/[-_.~0-9a-zA-Z]/) ? c : encodeURIComponent(c)))
+    .join('');
+}
+
+export function encodeWithoutSlashes(str: string): string {
+  return str
+    .split('')
+    .map(c => (c.match(/[-_.~0-9a-zA-Z/]/) ? c : encodeURIComponent(c)))
+    .join('');
+}
+
+function escapeRegExp(str: string) {
+  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+}
+
+export function applyPattern(
+  pattern: string,
+  fieldValue: string
+): string | undefined {
+  if (!pattern || pattern === '*') {
+    return encodeWithSlashes(fieldValue);
+  }
+
+  if (!pattern.includes('*') && pattern !== fieldValue) {
+    return undefined;
+  }
+
+  // since we're converting the pattern to a regex, make necessary precautions:
+  const regex = new RegExp(
+    '^' +
+      escapeRegExp(pattern)
+        .replace(/\\\*\\\*/g, '(.+)')
+        .replace(/\\\*/g, '([^/]+)') +
+      '$'
+  );
+
+  if (!fieldValue.match(regex)) {
+    return undefined;
+  }
+
+  return encodeWithoutSlashes(fieldValue);
+}
+
+interface MatchResult {
+  matchedFields: string[];
+  url: string;
+}
+
+export function match(
+  request: RequestType,
+  pattern: string
+): MatchResult | undefined {
+  let url = pattern;
+  const matchedFields = [];
+  for (;;) {
+    const match = url.match(/^(.*)\{([^}=]+)(?:=([^}]*))?\}(.*)/);
+    if (!match) {
+      break;
+    }
+    const [, before, field, pattern, after] = match;
+    matchedFields.push(field);
+    const fieldValue = getField(request, field);
+    if (typeof fieldValue === 'undefined') {
+      return undefined;
+    }
+    const appliedPattern = applyPattern(pattern, fieldValue.toString());
+    if (typeof appliedPattern === 'undefined') {
+      return undefined;
+    }
+    url = before + appliedPattern + after;
+  }
+
+  return {matchedFields, url};
+}
+
+export function flattenObject(request: RequestType): RequestType {
+  const result: RequestType = {};
+  for (const key in request) {
+    if (typeof request[key] === 'undefined') {
+      continue;
+    }
+
+    if (Array.isArray(request[key])) {
+      // According to the http.proto comments, a repeated field may only
+      // contain primitive types, so no extra recursion here.
+      result[key] = request[key];
+      continue;
+    }
+
+    if (typeof request[key] === 'object' && request[key] !== null) {
+      const nested = flattenObject(request[key] as RequestType);
+      for (const nestedKey in nested) {
+        result[`${key}.${nestedKey}`] = nested[nestedKey];
+      }
+      continue;
+    }
+
+    result[key] = request[key];
+  }
+
+  return result;
+}
+
+function camelToSnakeCase(str: string) {
```

Comment:
would it be worth pulling some of these helpers into a `utils/string.js`?

Reply:
Done!  (just `src/util.ts` for now)